blockchain/elrondBaseProxy.go

```diff
@@ -0,0 +1,265 @@
+	if proxy.sinceTimeHandler(proxy.lastFetchedTime) > proxy.cacheExpiryDuration {
+		return nil
+	}
+
+	return proxy.fetchedConfigs
+}
+
+func (proxy *elrondBaseProxy) cacheConfigs(ctx context.Context) (*data.NetworkConfig, error) {
+	proxy.mut.Lock()
+	defer proxy.mut.Unlock()
+
+	// maybe another parallel running go routine already did the fetching
+	cachedConfig := proxy.getCachedConfigs()
+	if cachedConfig != nil {
+		return cachedConfig, nil
+	}
+
+	log.Debug("Network config not cached. caching...")
+	configs, err := proxy.getNetworkConfigFromSource(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	proxy.lastFetchedTime = time.Now()
+	proxy.fetchedConfigs = configs
+
+	return configs, nil
+}
+
+// GetNetworkConfig retrieves the network configuration from the proxy
+func (proxy *elrondBaseProxy) getNetworkConfigFromSource(ctx context.Context) (*data.NetworkConfig, error) {
+	buff, err := proxy.GetHTTP(ctx, networkConfigEndpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	response := &data.NetworkConfigResponse{}
+	err = json.Unmarshal(buff, response)
+	if err != nil {
+		return nil, err
+	}
+	if response.Error != "" {
+		return nil, errors.New(response.Error)
+	}
+
+	return response.Data.Config, nil
+}
+
+// CheckShardFinalization will query the proxy and check if the target shard ID has a current nonce close to the cross
+// check nonce from the metachain
+// nonce(target shard ID notarized by meta) + maxNoncesDelta >= nonce(target shard ID)
+func (proxy *elrondBaseProxy) CheckShardFinalization(ctx context.Context, targetShardID uint32, maxNoncesDelta uint64) error {
+	if maxNoncesDelta < minAllowedDeltaToFinal {
+		return fmt.Errorf("%w, provided: %d, minimum: %d", ErrInvalidAllowedDeltaToFinal, maxNoncesDelta, minAllowedDeltaToFinal)
+	}
+	if targetShardID == elrondCore.MetachainShardId {
+		// we consider this final since the minAllowedDeltaToFinal is 1
+		return nil
+	}
+
+	nonceFromMeta, nonceFromShard, err := proxy.getNoncesFromMetaAndShard(ctx, targetShardID)
+	if err != nil {
+		return err
+	}
+
+	if nonceFromShard < nonceFromMeta {
+		return fmt.Errorf("shardID %d is syncing, meta cross check nonce is %d, current nonce is %d, max delta: %d",
+			targetShardID, nonceFromMeta, nonceFromShard, maxNoncesDelta)
+	}
+	if nonceFromMeta+maxNoncesDelta >= nonceFromShard {
+		return nil
+	}
+
+	return fmt.Errorf("shardID %d is stuck, meta cross check nonce is %d, current nonce is %d, max delta: %d",
+		targetShardID, nonceFromMeta, nonceFromShard, maxNoncesDelta)
+}
+
+func (proxy *elrondBaseProxy) getNoncesFromMetaAndShard(ctx context.Context, targetShardID uint32) (uint64, uint64, error) {
+	networkStatusMeta, err := proxy.GetNetworkStatus(ctx, elrondCore.MetachainShardId)
+	if err != nil {
+		return 0, 0, err
+	}
+
+	crossCheckValue := networkStatusMeta.CrossCheckBlockHeight
+	nonceFromMeta, err := extractNonceOfShardID(crossCheckValue, targetShardID)
+	if err != nil {
```

Comment:
A metachain node can provide the shard's nonce that it has notarized. 
For example, when we query a metachain node, it might return a string like this: `0: 9169897, 1: 9166353, 2: 9170524, ` which means that the metachain notarized from shard 0 the block with nonce 9169897, from shard 1 nonce 9166353 and from shard 2 the nonce 9170524. This means that the metablock that notarized these shard blocks when it will become final (usually in the next round), the shard nonces will become final. This translates in the fact that shards can not rollback the blocks behind those nonces no-matter-what.